### PR TITLE
TandemFoil: 3L/256d + lr=1e-4 + gc=0.5 — compound best insights

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Compound the three winning modifications: width scaling (3L/256d), lower LR (1e-4), and gc=0.5. Also probe 3L/320d as an even wider variant.

## Baseline

val_primary/surface_pressure_mae = **30.10** (PR #2810)

## Runs

Run all from `cd target/icml2026`.

**Run 1** (3L/256d + lr=1e-4 + gc=0.5, `CUDA_VISIBLE_DEVICES=0,1,2`):
```bash
SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset tandemfoil --optimizer lion --lr 1e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 256 --model-heads 4 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --wandb-group tandem-3L256d-lr1e4 --wandb-name tandem-3L256d-lr1e4-gc05
```

**Run 2** (3L/256d + lr=1e-4 + gc=1.0, gc comparison, `CUDA_VISIBLE_DEVICES=3,4,5`):
Same as Run 1 but `--grad-clip 1.0 --wandb-name tandem-3L256d-lr1e4-gc10`

**Run 3** (3L/320d + lr=1.25e-4 + gc=0.5 — wider, `CUDA_VISIBLE_DEVICES=6,7`):
Same as Run 1 but `--model-hidden-dim 320 --model-heads 5 --lr 1.25e-4 --grad-clip 0.5 --wandb-name tandem-3L320d-lr125e4`

## Key Insight

3L/256d uses `--model-hidden-dim 256 --model-heads 4`. 3L/320d uses `--model-hidden-dim 320 --model-heads 5`. This compounds three orthogonal improvements (width + lower LR + tighter gc) that individually each pushed toward 30.10. If they are genuinely orthogonal, compounding them should push further below that floor.

## Expected Outcome

val_primary/surface_pressure_mae < 30.10

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT tanjiro:

## Results

### Run 1: 3L/256d + lr=1e-4 + gc=0.5 (BEST)

| Metric | Value |
|--------|-------|
| **val_primary/surface_pressure_mae (best)** | **30.23** (epoch 121) |
| val_primary/surface_pressure_mae (terminal) | 31.24 (epoch 127) |
| test_primary/surface_pressure_mae | 30.68 (terminal checkpoint) |
| test_single_in_dist | 25.57 |
| test_geom_camber_cruise | 32.99 |
| test_geom_camber_rc | 39.48 |
| test_re_rand | 24.70 |
| Total epochs | 127 (180-min budget, ~1.42 min/epoch) |
| Peak GPU memory | ~84 GB (RTX PRO 6000 Blackwell) |
| W&B run | `ieqx2sfz` |

**Command:**
```bash
CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset tandemfoil --optimizer lion --lr 1e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 256 --model-heads 4 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --agent tanjiro --wandb-group tandem-3L256d-lr1e4 --wandb-name "tanjiro/tandem-3L256d-lr1e4-gc05"
```

### Run 2: 3L/256d + lr=1e-4 + gc=1.0

| Metric | Value |
|--------|-------|
| **val_primary/surface_pressure_mae (best)** | **32.10** (epoch 119) |
| val_primary/surface_pressure_mae (terminal) | 32.53 (epoch 126) |
| test_primary/surface_pressure_mae | 32.63 (terminal checkpoint) |
| test_single_in_dist | 27.69 |
| test_geom_camber_cruise | 36.19 |
| test_geom_camber_rc | 39.82 |
| test_re_rand | 26.84 |
| Total epochs | 126 (180-min budget, ~1.44 min/epoch) |
| Peak GPU memory | ~35 GB |
| W&B run | `hnxb6b2e` |

**Command:** Same as Run 1 but `--grad-clip 1.0 --wandb-name "tanjiro/tandem-3L256d-lr1e4-gc10"`

### Run 3: 3L/320d + lr=1.25e-4 + gc=0.5

| Metric | Value |
|--------|-------|
| **val_primary/surface_pressure_mae (best)** | **33.50** (epoch 100) |
| val_primary/surface_pressure_mae (terminal) | 33.79 (epoch 101) |
| test_primary/surface_pressure_mae | 34.45 (terminal checkpoint) |
| test_single_in_dist | 32.99 |
| test_geom_camber_cruise | 36.29 |
| test_geom_camber_rc | 40.99 |
| test_re_rand | 27.52 |
| Total epochs | 101 (180-min budget, ~1.79 min/epoch) |
| Peak GPU memory | ~83 GB |
| W&B run | `qf2oyd14` |

**Command:** Same as Run 1 but `--model-hidden-dim 320 --model-heads 5 --lr 1.25e-4 --wandb-name "tanjiro/tandem-3L320d-lr125e4-gc05"`

### Comparison vs Baseline (PR #2810: 3L/192d, lr=1.25e-4, gc=1.0 → 30.10)

| Run | Best val MAE | vs Baseline | Epochs |
|-----|-------------|-------------|--------|
| **Run 1** (256d, gc=0.5) | **30.23** | **+0.43%** | 127 |
| Run 2 (256d, gc=1.0) | 32.10 | +6.6% | 126 |
| Run 3 (320d, gc=0.5) | 33.50 | +11.3% | 101 |
| Baseline (192d, gc=1.0) | 30.10 | — | 157 |

### What happened

**Run 1 (gc=0.5) nearly matched the baseline** — 30.23 vs 30.10, only 0.43% higher. This is a strong result, but the compound of 3 improvements did not beat the simpler 192d configuration. The key insight:

1. **gc=0.5 decisively outperforms gc=1.0 at 256d width.** 30.23 vs 32.10 (6.2% gap). The tighter gradient clipping finds significantly deeper cosine troughs throughout training.

2. **256d width didn't pay off vs 192d.** The 256d model runs at ~1.42 min/epoch vs ~1.15 min/epoch for 192d (24% slower). This epoch penalty means 127 vs 157 epochs in the same wall-clock budget. The extra capacity does NOT compensate — the 192d model reaches the same MAE territory with more optimization steps.

3. **320d was disadvantaged by epoch throughput.** Only 101 epochs (1.79 min/epoch). Its best of 33.50 was still descending — with more time it would likely approach the 256d results, but the throughput cost is too high for a 180-min budget.

4. **All runs still descending at timeout.** Run 1's trough envelope: 35.88 (ep52) → 34.19 (ep68) → 33.85 (ep89) → 32.04 (ep98) → 31.43 (ep102) → 30.92 (ep115) → 30.23 (ep121). With more time, sub-30 is plausible.

5. **lr=1e-4 + gc=0.5 at 192d could be the next breakout.** The gc=0.5 advantage is clear, and lr=1e-4 works well. Compounding these at the faster 192d throughput (getting ~157+ epochs) could beat 30.10.

### Suggested follow-ups

- **lr=1e-4 + gc=0.5 at 3L/192d** — the winning combination from this experiment at the throughput-optimal architecture. More epochs should push deeper.
- **Longer budget for 3L/256d gc=0.5** — Run 1 was still descending. A 360-min budget could reach sub-30.
- **Best-checkpoint saving** — test metrics are from terminal (not best) checkpoint. Implementing save/load of best checkpoint would give accurate test metrics.

**Comment by tcapelle** [comment]
ADVISOR: Your Run 1 (3L/256d + lr=1e-4 + gc=0.5) reached 30.23 — tantalizingly close to the 30.10 baseline. Two key insights from your data:

1. **gc=0.5 is clearly superior to gc=1.0 at this config** (30.23 vs 32.10 — 6.2% gap)
2. **256d is too slow for 180-min budget** (127 epochs vs 157 at 192d)

The fix: try 3L/192d (the throughput-optimal width) with your winning hyperparameters:

**Run 1 — 3L/192d + lr=1e-4 + gc=0.5:**
```bash
cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion   --lr 1e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2   --no-use-ema --model-slices 64   --model-layers 3 --model-hidden-dim 192 --model-heads 3   --enable-fourier --enable-te-coord-frame --enable-cp-panel   --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction   --enable-pressure-prior-addition --epochs 999   --wandb-group tanjiro-192d-gc05
```

**Run 2 — 3L/192d + lr=1.25e-4 + gc=0.5 (isolate gc effect at baseline LR):**
```bash
cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion   --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2   --no-use-ema --model-slices 64   --model-layers 3 --model-hidden-dim 192 --model-heads 3   --enable-fourier --enable-te-coord-frame --enable-cp-panel   --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction   --enable-pressure-prior-addition --epochs 999   --wandb-group tanjiro-192d-gc05
```

This keeps your two confirmed improvements (gc=0.5, lr=1e-4) while removing the epoch-budget bottleneck. With 157+ epochs at 192d, you should clear the 30.10 barrier. Report val_primary/surface_pressure_mae.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2842 because ADVISOR: Closing this PR. TandemFoil is now at val=22.537 (#2924 MERGED). The 3L/256d+lr=1e-4+gc=0.5 experiments here (best ~30.23) are obsolete — 192d width at Lion+EMA+gc=0.5 is the winning recipe. Closing as superseded. Student will receive fresh assignment..
